File: 백준/Gold/2638. 치즈/치즈.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {

	static int n;
	static int m;

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		n = sc.nextInt();
		m = sc.nextInt();

		int[][] grid = new int[n + 1][m + 1];

		for (int i = 1; i < n + 1; i++) {
			for (int j = 1; j < m + 1; j++) {
				int tmp = sc.nextInt();
				if (tmp > 0) {
					grid[i][j] = tmp + 1;
				} else {
					grid[i][j] = tmp;
				}
			}
		}
		
//		for(int i = 0; i<n+1; i++) {
//			for(int j = 0; j<m+1;j++) {
//				System.out.print(grid[i][j]+" ");
//			}
//			System.out.println();
//		}
//		System.out.println("=================");
		
		int cnt = 0;
		while (true) {
			visited = new boolean[n + 1][m + 1];
			bfs(grid, 0, 0);
			fill2(grid);
			if(exit == 0) {
				break;
			}
			cnt++;
			
		}
		System.out.println(cnt);

	}

	static int[] row = { 1, -1, 0, 0 };
	static int[] col = { 0, 0, 1, -1 };
	
	static int exit;

	static boolean[][] visited;

	private static void bfs(int[][] grid, int r, int c) {
		exit = 0;
		visited[r][c] = true;

		Queue<Integer> q = new LinkedList<>();

		q.offer(r);
		q.offer(c);

		while (!q.isEmpty()) {

			int rnode = q.poll();
			int cnode = q.poll();

			for (int k = 0; k < 4; k++) {
				int nr = rnode + row[k];
				int nc = cnode + col[k];

				if (nr >= 0 && nr < n+1 && nc >= 0 && nc < m+1 &&!visited[nr][nc]&& grid[nr][nc] == 0) {
					visited[nr][nc] = true;
					q.offer(nr);
					q.offer(nc);
				} else if (nr >= 0 && nr < n+1 && nc >= 0 && nc < m+1 && grid[nr][nc] > 0) {
					exit++;
					visited[nr][nc] = true;
					grid[nr][nc]--;
					if (grid[nr][nc] < 0) {
						grid[nr][nc] = 0;
					}
				}

			}

		}
//		for(int i = 0; i<n+1; i++) {
//			for(int j = 0; j<m+1;j++) {
//				System.out.print(grid[i][j]+" ");
//			}
//			System.out.println();
//		}
//		System.out.println("=================");

	}
	
	private static void fill2(int[][] grid) {
		for(int i = 0; i<n;i++) {
			for(int j = 0; j<m;j++) {
				if(grid[i][j]>0) {
					grid[i][j]=2;
				}
			}
		}
		
	}
}

File: 백준/Gold/2638. 치즈/치즈.py
'''
구상
카운트를 남겨서 진행
'''
from collections import deque

n,m = map(int, input().split())
grid = [list(map(int, input().split())) for i in range(n)]
time = 0

def all_melting():
    for i in range(n):
        for j in range(m):
            if grid[i][j]==1:
                return False
    return True

row = [-1,1,0,0]
col = [0,0,1,-1]

def bfs(sr,sc):
    visited = [[False]*m for i in range(n)]
    visited[sr][sc] = True
    q = deque([(sr,sc)])
    counting_grid = [[0]*m for i in range(n)]

    while q:
        r, c = q.popleft()

        for k in range(4):
            nr = r+row[k]
            nc = c+col[k]
            if not(0<=nr<n and 0<=nc<m):
                continue
            if not visited[nr][nc] and grid[nr][nc]==0:
                q.append((nr,nc))
                visited[nr][nc] = True
            if grid[nr][nc]==1:
                counting_grid[nr][nc] +=1
    # print("--------------")
    # for _ in counting_grid:
    #     print(*_)

    # grid 가 1 이고 count_grid가 2이상인 애들은 녹임
    for i in range(n):
        for j in range(m):
            if grid[i][j]==1 and counting_grid[i][j]>=2:
                grid[i][j]=0
    # print("--------------")
    # for _ in grid:
    #     print(*_)

while True:
    if(all_melting()):
        break
    time+=1
    bfs(0,0)
print(time)